torch/csrc/jit/codegen/cuda/kernel_db/kernel_db.cpp

```diff
@@ -0,0 +1,253 @@
+
+  return get(
+      kernel_db_dir,
+      kernel_db_file,
+      true,
+      !isOptionEnabled(EnableOption::KernelDb),
+      false);
+}
+
+KernelDb& KernelDb::get(
+    const std::string& kernel_db_dir,
+    const std::string& kernel_db_file,
+    bool use_temp_dir,
+    bool disabled,
+    bool reset) {
+  std::lock_guard<std::mutex> guard(kernel_db_lock);
+
+  // The KernelDb is minimally constructed to at least hold the disable and
+  // initialized booleans
+  static KernelDb singleton(disabled);
+
+  if (reset) {
+    singleton.disabled_ = disabled;
+    singleton.initialized_ = false;
+    singleton.kernel_map_.clear();
+    singleton.kernel_db_path_.clear();
+    singleton.kernel_db_txt_file_.clear();
+  }
+
+  // Intialize the Db if it isn't already disabled
+  if (!singleton.disabled_ && !singleton.initialized_) {
+    // If the appropriate files are not found or unable to be created, disable
+    if (!singleton.open(kernel_db_dir, kernel_db_file, use_temp_dir)) {
+      singleton.disabled_ = true;
+    } else {
+      singleton.initialized_ = true;
+    }
+  }
+  return singleton;
+}
+
+bool KernelDb::query(
+    const std::string& kernel_code,
+    const std::string& compile_args,
+    std::string& kernel_name,
+    std::vector<char>& cubin) const {
+  FUSER_PERF_SCOPE("KernelDb::query");
+  std::lock_guard<std::mutex> guard(kernel_db_lock);
+  bool status = false;
+  auto db_entry = kernel_map_.find(kernel_code);
+
+  // Kernel Match is found
+  if (db_entry != kernel_map_.end()) {
+    // Make sure the compilation args also match
+    if (db_entry->second.compile_args == compile_args) {
+      // Copy the cubin to a data buffer and record the kernel name for module
+      // loading
+      fs::path cubin_file_path = kernel_db_path_ / db_entry->second.cubin_file;
+      if (copy_from_binary_file(cubin_file_path.string(), cubin)) {
+        kernel_name = db_entry->second.kernel_name;
+        status = true;
+      }
+    }
+  }
+  return status;
+}
+
+// This method will write a cubin and the kernel code to files as well as add
+// an entry to the db csv file.
+bool KernelDb::write(
+    const std::string& kernel_code,
+    const std::string& compile_args,
+    const std::string& kernel_name,
+    const std::vector<char>& cubin) {
+  FUSER_PERF_SCOPE("KernelDb::write");
+  std::lock_guard<std::mutex> guard(kernel_db_lock);
+  bool status = false;
+
+  // If the kernel doesn't already exist in the hash map, add it.
+  if (kernel_map_.count(kernel_code) == 0) {
+    // The cubin and kernel code files are given a unique number based on the
+    // size of the hash map.
+    std::string kernel_num =
+        std::to_string(static_cast<unsigned long>(kernel_map_.size()));
+
+    // Kernel Code File path
+    std::string code_file_name("kernel_" + kernel_num + ".cu");
+    fs::path code_file_path = kernel_db_path_ / code_file_name;
+
+    // Cubin File path
+    std::string cubin_file_name("kernel_" + kernel_num + ".cubin");
```

Comment:
technically we are writing ptx for pre CUDA 11.1
Not sure if that's relevant any more.

Reply:
It shouldn't matter except that the `.cubin` file suffix would be technically wrong.